scheduler/scheduler.go

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2021 The Gort Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package scheduler
+
+import (
+	"context"
+	"time"
+
+	"github.com/getgort/gort/adapter"
+	"github.com/getgort/gort/command"
+	"github.com/getgort/gort/data"
+	"github.com/getgort/gort/dataaccess"
+	"github.com/getgort/gort/telemetry"
+	"github.com/go-co-op/gocron"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/codes"
+)
+
+type CommandScheduler struct {
+	cron     *gocron.Scheduler
+	Commands chan data.CommandRequest
+	da       dataaccess.DataAccess
+}
+
+func NewCommandScheduler(da dataaccess.DataAccess) CommandScheduler {
+	cron := gocron.NewScheduler(time.Local)
+	cron.TagsUnique()
+	return CommandScheduler{
+		cron:     cron,
+		Commands: make(chan data.CommandRequest),
+		da:       da,
+	}
+}
+
+func (cs *CommandScheduler) Start() {
+	cs.cron.StartAsync()
+}
+
+func (cs *CommandScheduler) schedule(ctx context.Context, command data.ScheduledCommand) error {
+	err := cs.da.ScheduleCreate(ctx, &command)
+	if err != nil {
+		return err
+	}
+	_, err = cs.cron.Cron(command.Cron).Do(func() { //todo tag with scheduleid
+		tr := otel.GetTracerProvider().Tracer(telemetry.ServiceName)
+		ctx, sp := tr.Start(context.Background(), "adapter.handleIncomingEvent")
+		defer sp.End()
+
+		req := data.CommandRequest{
+			CommandEntry: command.CommandEntry,
+			Adapter:      command.Adapter,
+			ChannelID:    command.ChannelID,
+			Context:      ctx,
+			Parameters:   command.Parameters,
+			Timestamp:    time.Now(),
+			UserID:       command.UserID,
+			UserEmail:    command.UserEmail,
+			UserName:     command.UserName,
+		}
+
+		err := cs.da.RequestBegin(ctx, &req)
+		if err != nil {
+			sp.RecordError(err)
+			sp.SetStatus(codes.Error, "Failed to begin request")
+			return
+		}
+
+		cs.Commands <- req
+	})
+
+	return err
+}
+
+func (cs *CommandScheduler) Add(ctx context.Context, cron string, cmdEntry data.CommandEntry, parameters data.CommandParameters, userID, userEmail, userName, adapterName, channelID string) error {
```

Comment:
With a parameter list like this, you may as well just accept a `data.ScheduledCommand` value